### ISSUE TITLE
On macOS, crash calling FlutterViewController.init(FlutterDartProject)

### ISSUE THREAD

**Issue by fertrig** [flutter/flutter#124829]
This crash happens on latest flutter beta `3.10.0-1.1.pre` on macOS.

The error message is: 

```
blue_macos[9882:79310701] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x6000017348d0
blue_macos[9882:79310701] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x6000017348d0
```

### Steps to reproduce

(I also created a repo that can reproduce the crash. See: https://github.com/fertrig/flutter_view_controller_crash)

On macOS:

1. Set up a working directory to host a flutter project, a flutter bundle and a macos project:
```
mkdir flutter_view_controller_crash
mkdir flutter_view_controller_crash/fl_beta
cd flutter_view_controller_crash/fl_beta
// source flutter beta `3.10.0-1.1.pre`
flutter create blue
```

2. Create a flutter bundle from the project you just created
```
mkdir flutter_view_controller_crash/blue_bundle
cd flutter_view_controller/fl_beta/blue
flutter build bundle -t lib/main.dart --debug --target-platform darwin --asset-dir ../blue_bundle/flutter_assets
```

3. Create new Xcode project with these details:
    - Platform: macOS
    - Application: App
    - Product Name: blue_macos
    - Interface: Storyboard
    - Language: Swift
    - Create it inside fl_beta directory

4. Once the xcode project is created select blue_macos on the left, then Targets > blue_macos > Signing & Capabilities. Then delete App Sandbox and Hardened Runtime.

5. Copy the `FlutterMacOS.framework` to blue_macos directory:
```
cp -R /path/to/flutter-beta-sdk/bin/cache/artifacts/engine/darwin-x64/FlutterMacOS.framework /path/to/flutter_view_controller_crash/fl_beta/blue_macos/
```

6.  In the xcode project, go to Targets > blue_macos> General > Framework, Libraries and Embedded Content. Add FlutterMacOS.framework from fl_beta/blue_macos/FlutterMacOS.framework

7. Replace the contents of AppDelegate.swift with code that uses the bundle to create a `FlutterDartProject`, which then calls `FlutterViewController.init(FlutterDartProject)`:
```swift
import Cocoa
import FlutterMacOS

@main
class AppDelegate: NSObject, NSApplicationDelegate {

    func applicationDidFinishLaunching(_ aNotification: Notification) {
        _launch_1()
    }

    func applicationWillTerminate(_ aNotification: Notification) {
    }

    func applicationSupportsSecureRestorableState(_ app: NSApplication) -> Bool {
        return true
    }

    func _launch_1() {
        let bundleUrl = URL(fileURLWithPath: "/PATH/TO/blue_bundle") // replace with your path
        let bundle = Bundle.init(path: bundleUrl.path)
        let project = FlutterDartProject.init(precompiledDartBundle: bundle) 
        let controller = FlutterViewController.init(project: project) // set your breakpoint here
        let window = NSWindow()
        window.contentViewController = controller
        let windowController = NSWindowController()
        windowController.contentViewController = window.contentViewController
        
        window.setContentSize(NSSize(width: 300, height: 400))
        window.setFrameOrigin(NSPoint(x: 100.0, y: 100.0))
        window.title = "blue"
        
        windowController.window = window
        windowController.showWindow(self)
    }

}

```

8. Run the app from xcode, and notice the error message:
```
blue_macos[9882:79310701] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x6000017348d0
blue_macos[9882:79310701] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x6000017348d0
```

The error message is emitted when calling `FlutterViewController.init(project: project)` above. It also crashes the control flow. The flutter window does not show.

If you run the same steps with flutter stable, it doesn't crash, and the flutter window shows with the flutter demo app.

I created this repo if you need help setting things up: https://github.com/fertrig/flutter_view_controller_crash

<details><summary>flutter doctor -v</summary>
<p>

```
[✓] Flutter (Channel beta, 3.10.0-1.1.pre, on macOS 12.6.2 21G320 darwin-x64, locale en-US)
    • Flutter version 3.10.0-1.1.pre on channel beta at /Users/fertrig/development/flutter-sdks/beta-channel
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ecdb7841df (2 days ago), 2023-04-11 20:47:40 -0700
    • Engine revision 870e640e7e
    • Dart version 3.0.0 (build 3.0.0-417.1.beta)
    • DevTools version 2.23.1

[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/fertrig/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[!] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    ! CocoaPods 1.10.0 out of date (1.11.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.76.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.60.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.6.2 21G320 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 112.0.5615.49

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.

```

</p>
</details>

**Comment by huycozy** [comment]
Hi @fertrig 
Could you share the complete output? 

This is the output from my end:

```
2023-04-14 15:10:54.955252+0700 blue_macos[2352:29844] Cannot find executable for CFBundle 0x7fb75bc16a70 </Users/huynq/Documents/NEVERCODE.IO/RepoduceProjects/OPClone/flutter_view_controller_crash/fl_beta/blue_bundle> (not loaded)
2023-04-14 15:10:54.955784+0700 blue_macos[2352:29844] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x600003e2ca00
2023-04-14 15:10:54.956285+0700 blue_macos[2352:29844] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x600003e2ca00
```

Even though I see `blue_bundle/flutter_assets` directory exists on the path but it still has that error. Do you see the same error? 

Also, the window can be shown from my end: 
![Screenshot 2023-04-14 at 15 16 01](https://user-images.githubusercontent.com/104349824/231986061-9bc59322-f64e-4797-a4e4-15fdcbcd7777.png)

**Comment by huycozy** [comment]
Also, can you please retry this on the latest master channel to see if the issue still occurs or not?

**Comment by fertrig** [comment]
Hi @huycozy,

### Full output
I get the same output you get. The first error is innocuous. It has been there for a while now. If you run it using flutter stable you will also get the first error which doesn't cause any side effects for the user.

Here is the full output I get:
```
2023-04-14 17:23:20.275899-0400 blue_macos[7980:80326741] Cannot find executable for CFBundle 0x7f9cf260dda0 </Users/fertrig/development/explorations/flutter_view_controller_crash/fl_beta/blue_bundle> (not loaded)
2023-04-14 17:23:20.276380-0400 blue_macos[7980:80326741] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x600000f141b0
2023-04-14 17:23:20.276819-0400 blue_macos[7980:80326741] -[blue_macos.AppDelegate setTerminationHandler:]: unrecognized selector sent to instance 0x600000f141b0
```

### Window shown
The window you see is the one from the storyboard. It is not the one we are launching via code. If you run the sample project using flutter stable you will see two windows: the storyboard one which is blank, and the flutter window we launch programmatically with the contents of the Flutter Demo Home Page.

### Error still occurs in master
I tried it on the latest master `Flutter 3.10.0-6.0.pre.11` and the error still occurs. I updated the repo for this crash with a project for flutter master: https://github.com/fertrig/flutter_view_controller_crash

To summarize, the crash occurs on latest beta and latest master. It doesn't crash on latest stable.

**Comment by huycozy** [comment]
Thanks for your response. I can see the difference between Flutter channels as you described. 

| stable | beta |
| --------------- | --------------- |
<img width="400" src="https://user-images.githubusercontent.com/104349824/232378343-7ca2371a-2411-4c60-acdf-79e00244acd0.png"> | <img width="400" src="https://user-images.githubusercontent.com/104349824/232378355-45bc000e-5b1d-4431-997c-8a6a054c2173.png">

<details>
<summary>flutter doctor -v (stable, beta, master)</summary>

```bash
[✓] Flutter (Channel stable, 3.7.11, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.7.11 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f72efea43c (31 hours ago), 2023-04-11 11:57:21 -0700
    • Engine revision 1a65d409c7
    • Dart version 2.19.6
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.77.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 112.0.5615.49

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.10.0-6.0.pre.28, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.10.0-6.0.pre.28 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 50171bbae3 (10 hours ago), 2023-04-16 13:37:40 -0400
    • Engine revision 4a998b73a2
    • Dart version 3.1.0 (build 3.1.0-11.0.dev)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.77.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 112.0.5615.121

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

```bash
[!] Flutter (Channel beta, 3.10.0-1.1.pre, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.10.0-1.1.pre on channel beta at /Users/huynq/Documents/GitHub/flutter_beta
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_beta. Consider adding /Users/huynq/Documents/GitHub/flutter_beta/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_beta. Consider adding /Users/huynq/Documents/GitHub/flutter_beta/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ecdb7841df (5 days ago), 2023-04-11 20:47:40 -0700
    • Engine revision 870e640e7e
    • Dart version 3.0.0 (build 3.0.0-417.1.beta)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] IntelliJ IDEA Community Edition (version 2022.1.1)
    • IntelliJ at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/221.5591.52/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.77.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 112.0.5615.121

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Comment by gspencergoog** [comment]
Hmm.  Maybe.

Is this app using it's own `NSApplicationDelegate` that isn't a subclass of `FlutterAppDelegate`? If so, that would explain the lack of a setter for the `terminationHandler`.

**Comment by fertrig** [comment]
@gspencergoog Yes, it is using its own `NSApplicationDelegate`. Is there a way to set the `terminationHandler` or a workaround?

**Comment by gspencergoog** [comment]
It would need to derive from `FlutterAppDelegate`, or implement its own termination handler property. It could also have its own reference to a FlutterAppDelegate and forward its own `terminationHandler` accessor to that copy.

**Comment by gspencergoog** [comment]
I'll see if I can add a `respondsToSelector` check before calling it so that it won't crash in that case, and just log.

**Comment by fertrig** [comment]
@gspencergoog Changing our code to derive from `FlutterAppDelegate` worked for us. However, my first approach was to implement a termination handler property in our custom app delegate, which I'm not sure it is possible.

The type `FlutterEngineTerminationHandler` is internal. The `terminationHandler` property of `FlutterEngine` is also internal. So I can't use them.

Also, the docs in `shell/platform/darwin/macos/framework/Headers/FlutterAppDelegate.h` state:
```
 * App delegates for Flutter applications are *not* required to inherit from
 * this class. Developers of custom app delegate classes should copy and paste
 * code as necessary from FlutterAppDelegate.mm.
```
I'm not sure about that comment. FlutterEngine.mm references FlutterAppDelegate. Thus, it seems a developer of a custom app delegate would also need to implement its own FlutterEngine.

How could we implement our own custom app delegate?

For now, deriving from `FlutterAppDelegate` worked, even though we are not using its `mainFlutterWindow` property.

**Comment by gspencergoog** [comment]
@fertrig Thanks for the comments.  It sounds like we need to make  `FlutterEngineTerminationHandler`  and the `terminationHandler` property of `FlutterEngine` public in order to make this possible. We try and err on the side of making things internal because making them public makes it harder to change the interface as time goes on. I think we were assuming that people implementing a custom app delegate were probably also implementing their own engine, or at least subclassing the existing engine, which sounds like it's not a good assumption in your case.

Is there a reason why subclassing `FlutterAppDelegate` isn't sufficient/desirable for your use case?

**Comment by fertrig** [comment]
The purpose of the `FlutterAppDelegate` doesn't align well with our use case. The `FlutterAppDelegate` is a "subclass for simple apps that want default behavior." Our app is not simple nor wants the default behavior.

It happens to work for now. But it feels risky to use it.

Our application launches multiple windows. Each window with their own `FlutterViewController` and `FlutterDartProject` and their own flutter bundle. 

The `FlutterAppDelegate` has some default behaviors and a main window which we don't use. I think that may give us trouble in the future as the API changes.

Here is something that may work for us:

```
class BaseFlutterAppDelegate : NSObject, NSApplicationDelegate {
  var terminationHandler ...
}

class FlutterAppDelegate : BaseFlutterAppDelegate {
  var mianFlutterWindow ...
  var applicationMenu ...
}
```
Then, `FlutterEngine` would use `BaseFlutterAppDelegate` to set the termination handler. And, complex apps like ours, can subclass `BaseFlutterAppDelegate` without having to subclass `FlutterAppDelegate` and without having to implement our own engine.

**Comment by gspencergoog** [comment]
cc @cbracken

**Comment by helgoboss** [comment]
> I'll see if I can add a `respondsToSelector` check before calling it so that it won't crash in that case, and just log.

That would be great! I'm running into the same issue and just like @woolfman007 in https://github.com/flutter/flutter/issues/127476, I don't have control over the app delegate because the Flutter app will run as a plug-in (dylib) in a host that's completely unaware of Flutter. 

I guess this is the line which needs the `respondsToSelector` check:

https://github.com/flutter/engine/blob/0da06de991a90c169329d0d4788dcbcb16ebca0d/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm#L183

I think this change would also make sense in terms of architecture. Right now, the engine makes assumptions about the application in which it is running, which could be easily avoided by using `respondToSelector`.

BTW, on Windows, starting Flutter from a plug-in (DLL) is already possible and quite straightforward. Even embedding the Flutter window as a child of another window is a breeze.

**Comment by cbracken** [comment]
@gspencergoog is this fixed now?

**Comment by gspencergoog** [comment]
Yep
